### PR TITLE
fix: fix removeUnnecessaryJoints and combineSkeletons

### PR DESCRIPTION
**Pull request by 0b5vr**
This PR fixes `removeUnnecessaryJoints` and `combineSkeletons`.

It processes the same bone index attribute twice or more, which collapses the mesh.

### TODOs

- [x] fix combineSkeletons

### Points need review

- [ ] Is it working properly?

### CODE REVIEW

File: packages/three-vrm/src/VRMUtils/removeUnnecessaryJoints.ts [hunk 1]
```diff
@@ -46,64 +46,72 @@ export function removeUnnecessaryJoints(
...
-        bones[i] = bones[0];
-        boneInverses[i] = boneInverses[0];
-      }
+    const nBones = experimentalSameBoneCounts ? maxBones : newToOld.size;
+
+    for (let newIndex = 0; newIndex < nBones; newIndex++) {
+      const oldIndex = newToOld.get(newIndex) ?? 0;
+
+      bones.push(mesh.skeleton.bones[oldIndex]);
+      boneInverses.push(mesh.skeleton.boneInverses[oldIndex]);
```
Comment (0b5vr) [inline packages/three-vrm/src/VRMUtils/removeUnnecessaryJoints.ts:114]: Considering cases multiple meshes share the same skin index attribute but not the same bones (which is unlikely) or boneInverses (which is more likely), we moved these lines here

File: packages/three-vrm/src/VRMUtils/combineSkeletons.ts [hunk 2]
```diff
@@ -9,98 +9,193 @@ import * as THREE from 'three';
...
+  return skinnedMeshes;
+}
+
+/**
+ * List all skin indices used by the given geometry.
+ * If the skin weight is 0, the index won't be considered as used.
+ * @param skinIndexAttr The skin index attribute
+ * @param skinWeightAttr The skin weight attribute
+ */
+function listUsedIndices(geometry: THREE.BufferGeometry): Set<number> {
```
Comment (ke456-png) [inline packages/three-vrm/src/VRMUtils/combineSkeletons.ts]: nits: Please fix the `@param` comments.

File: packages/three-vrm/src/VRMUtils/combineSkeletons.ts [hunk 3]
```diff
@@ -9,98 +9,193 @@ import * as THREE from 'three';
...
-  const skinnedMeshes = new Set<THREE.SkinnedMesh>();
-  const geometryToSkinnedMesh = new Map<THREE.BufferGeometry, THREE.SkinnedMesh>();
+  const skinnedMeshes = collectSkinnedMeshes(root);
+
+  // List all bones and boneInverses for each meshes
+  const attributeUsedIndexSetMap = new Map<THREE.BufferAttribute | THREE.InterleavedBufferAttribute, Set<number>>();
+  const meshBoneInverseMapMap = new Map<THREE.SkinnedMesh, Map<THREE.Bone, THREE.Matrix4>>();
+  for (const mesh of skinnedMeshes) {
+    const boneInverseMap = listUsedBones(mesh, attributeUsedIndexSetMap);
+    meshBoneInverseMapMap.set(mesh, boneInverseMap);
```
Comment (ke456-png) [inline packages/three-vrm/src/VRMUtils/combineSkeletons.ts:28]: I think the side effects are not clear. How about setting meshBoneInverseMapMap within the function?

File: packages/three-vrm/src/VRMUtils/combineSkeletons.ts (hunk 2 above)
Comment (0b5vr) [reply packages/three-vrm/src/VRMUtils/combineSkeletons.ts]: oh no

File: packages/three-vrm/src/VRMUtils/combineSkeletons.ts (hunk 3 above)
Comment (0b5vr) [reply packages/three-vrm/src/VRMUtils/combineSkeletons.ts:28]: Thank you. I will rewrite to the more straightforward way without side effects.

**Review by ke456-png — APPROVED** [review]
Looks good.

**Pull request merged**